### ISSUE TITLE
Make resolve prototype method unenumerable

### ISSUE THREAD

**Issue by rigon** [rigon/hapi-url#1]
When you override the `resolve` prototype method, make it unenumerable via `Object.defineProperty` (via [Stackoverflow](https://stackoverflow.com/questions/31840286/how-to-get-the-full-url-for-a-request-in-hapi/49540619#comment86161776_49540619))

**Comment by gilly3** [comment]
When you add the `resolve` method, you change the interface of the `Url` instance, slightly.  Before the addition, the `resolve` method is in the prototype, so it doesn't show up in, eg, calls to `Object.keys(url)`.  After the addition, the `resolve` method is a direct property of the `Url` instance, and so it now is included in the return value from `Object.keys(url)`.  One way to resolve that is to make the property unenumerable.  [Enumerable properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) are those properties of an object that show up when enumerating an object via `for (var key in object) { ... }`.  To make a property unenumerable, instead of saying `url.resolve = resolveFn`, you would use `Object.defineProperty()`:

```javascript
Object.defineProperty(url, "resolve", {
  value: resolveFn,
  writable: true,
  configurable: true,
  enumerable: false
});
```

Definitely read [the documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty) for `Object.defineProperty()`.  Lots of good information there.

Upon further reflection, this is bad advice.  While making it unenumerable does prevent the property from appearing in `Object.keys(url)`, the interface is still changed and so it may not be the ideal solution.  The `resolve` function doesn't appear in `Object.keys(url)`, not because it's unenumerable, but because it's a [prototype method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain).  A better solution would be to add your resolve method to the prototype chain and leave it as enumerable (since the original is enumerable).  The easy way to do that would be to just override `Url.prototype.resolve`:

```javascript
Url.prototype.resolve = resolveFn;
```

But, that's no good, because it will apply to _all_ instances of `Url`, not just the ones you create.  The right way to do it would be to add your own link in the prototype chain.  You can do so by using [`Object.setPrototypeOf()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) and also [`Object.create()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create):

```javascript
const proto = Object.create(Url.prototype); // create new object with Url.prototype as its prototype
proto.resolve = resolveFn; // add a custom resolve function
Object.setPrototypeOf(url, proto);  // use proto as the prototype of url
```

In this way, all of the prototype methods from `Url` are preserved, but you are able to override the `resolve` method for _your_ instances of `Url` without overriding all other instances of `Url`.

**TL;DR**

Don't worry about making it unenumerable, instead add your own `resolve` method to the prototype chain.
